src/pint/toa.py

```diff
@@ -1491,31 +1532,41 @@ def get_dm_errors(self):
             raise AttributeError("No DM error is provided.")
         return np.array(result)[valid] * pint.dmu
 
-    def get_groups(self, gap_limit=None):
+    def get_groups(self, gap_limit=2 * u.h, add_column=False):
         """Flag toas within gap limit (default 2h = 0.0833d) of each other as the same group.
 
         Groups can be larger than the gap limit - if toas are separated by a gap larger than
         the gap limit, a new group starts and continues until another such gap is found.
 
-        Groups with a two-hour spacing are pre-computed when the TOAs object is constructed,
+        Groups with a two-hour spacing (default) are pre-computed when the TOAs object is constructed,
         and these can rapidly be retrieved from ``self.table`` (which this function will do).
 
         Parameters
         ----------
         gap_limit : :class:`astropy.units.Quantity`, optional
             The minimum size of gap to create a new group. Defaults to two hours.
+        add_column : bool, optional
+            Whether or not to add a `groups` column to the TOA table (default: False)
 
         Returns
         -------
         groups : array
             The group number associated to each TOA. Groups are numbered chronologically
             from zero.
         """
-        # TODO: make all values Quantity objects for consistency
-        if gap_limit is None:
-            gap_limit = 2 * u.h
-        if "groups" not in self.table.colnames or gap_limit != 2 * u.h:
-            return _group_by_gaps(self.get_mjds().value, gap_limit.to_value(u.d))
+        if (
+            ("groups" not in self.table.colnames)
+            or ("group_gap" not in self.table.meta)
+            or (gap_limit != self.table.meta["group_gap"])
+        ):
+            groups = _group_by_gaps(
+                self.get_mjds().to_value(u.d), gap_limit.to_value(u.d)
+            )
+            if add_column:
+                self.table.add_column(groups, name="groups")
+                self.table.meta["group_gap"] = gap_limit
```

Comment:
Does this metadata survive table subset selection (for example due to selecting a time range or TOA excision)? Should it?